### ISSUE TITLE
MolGIN paper implementation

### ISSUE THREAD

**Issue by thegodone** [pyg-team/pytorch_geometric#1729]
## 🚀 Feature

I am interesting to implement or see your advice on the following paper. Equation 5 the dimension of the torch objects in the addition between the (1+eps)*h and sum of concate(h,z) looks strange. Can you tell me if this is compatible to the pytorch-geometric package:

https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9189862

## Motivation

This new model looks better than GIN (to be confirmed)

**Comment by rusty1s** [comment]
That is indeed compatible, and can be implemented in the `message` function via `torch.cat`, *e.g.*:
```python
def message(self, h_j, z_j):
    return torch.cat([h_j, z_j], dim=-1)
```

**Comment by thegodone** [comment]
Thanks a lot and about equation 7, this sigmoid function:

    def forward(self, x, edge_index, edge_attr):
        """"""        
        x = x.unsqueeze(-1) if x.dim() == 1 else x        
        Out = self.nn((1 + self.eps) * x + self.propagate(edge_index, x=x, edge_attr=edge_attr))
        return out

    def message(self, x_j, edge_attr):
        return torch.cat(F.sigmoid(x_j), edge_attr], dim=-1)

correct ?

**Comment by rusty1s** [comment]
That looks correct to me.

**Comment by thegodone** [comment]
Try several possibilities around nothing work when I concate:

here my current version

`class MolGINConv(MessagePassing):
     """The modified :class:`MolGINConv` operator from the `"Enhanced Graph Isomorphism Network for
    Molecular ADMET Properties Prediction" https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9189862`_
    paper introduce the gate and the edge concate procedure => the emb_dim can change layer by layer ...
    case the gate need to be dynamic with the input size
    :param nn: (torch.nn.Module): A neural network :math:`h_{\mathbf{\Theta}}` that
            maps node features :obj:`x` of shape :obj:`[-1, in_channels]` to
            shape :obj:`[-1, out_channels]`, *e.g.*, defined by
            :class:`torch.nn.Sequential`.
    :param eps (float, optional): (Initial) :math:`\epsilon` value.
            (default: :obj:`0`)
    :param train_eps (bool, optional): If set to :obj:`True`, :math:`\epsilon`
            will be a trainable parameter. (default: :obj:`False`)
        **kwargs (optional): Additional arguments of
            :class:`torch_geometric.nn.conv.MessagePassing`.
    """
    def __init__(self, nn, eps=0, train_eps=False, **kwargs):
        super(MolGINConv, self).__init__(aggr='add', **kwargs)
        self.nn = nn
        self.initial_eps = eps
        emb_dim = 49
        if train_eps:
            self.eps = torch.nn.Parameter(torch.Tensor([eps]))
        else:
            self.register_buffer('eps', torch.Tensor([eps]))
        # self.edge_embedding1 = torch.nn.Embedding(num_bond_type, emb_dim)
        # adding the gate
        self.lin = torch.nn.Linear(emb_dim, emb_dim, bias=True) # this is not correct!
        self.act = torch.nn.Sigmoid()
        self.reset_parameters()

    def reset_parameters(self):
        reset(self.nn)
        # initialize the gate
        glorot(self.lin.weight)
        self.eps.data.fill_(self.initial_eps)


    def forward(self, x, edge_index, edge_attr):
        x = x.unsqueeze(-1) if x.dim() == 1 else x
        #edge_attr = self.edge_embedding1(edge_attr)
        out = self.nn((1 + self.eps) * x + self.propagate(edge_index, x=x, edge_attr=edge_attr))
        return out

    def message(self, x_j, edge_attr):
        # create the gate of x_j itself 
        gatex_j = x_j
        #gatex_j = self.act(self.lin(x_j))
        # compute the gate ie gatex_j*x_j + concate it with edge attribute 
        #return gatex_j*x_j
        return torch.cat([gatex_j*x_j, edge_attr], dim=-1)`

message works only if I do only one convolution using this code:
`    def message(self, x_j, edge_attr):
        # create the gate of x_j itself 
        gatex_j = self.act(self.lin(x_j))
        return gatex_j*x_j

with 2 convolutions:
/opt/conda/lib/python3.6/site-packages/torch/nn/functional.py in linear(input, weight, bias)
   1608     if input.dim() == 2 and bias is not None:
   1609         # fused op is marginally faster
-> 1610         ret = torch.addmm(bias, input, weight.t())
   1611     else:
   1612         output = input.matmul(weight.t())

RuntimeError: size mismatch, m1: [834 x 120], m2: [49 x 49] at /pytorch/aten/src/THC/generic/THCTensorMathBlas.cu:283

Now is I try to concate with this other code: 
`    def message(self, x_j, edge_attr):
        # create the gate of x_j itself 
        gatex_j = x_j
        # compute the gate ie gatex_j*x_j + concate it with edge attribute 
        return torch.cat([gatex_j*x_j, edge_attr], dim=-1)`

got this error 
--> 143         return torch.cat([gatex_j*x_j, edge_attr], dim=-1)
    144 
    145 

RuntimeError: Sizes of tensors must match except in dimension 0. Got 410 and 834
 
I play with ESOL dataset
Any solution or advice ?

**Comment by rusty1s** [comment]
Your first operator looks correct to me, but `emb_dim` should not be static (and actually depends on the output dimension of `self.nn` of the previous layer).

**Comment by thegodone** [comment]
Thanks Matthias,

ok but do you really think that equation 7 is correct based on dimensions
for the addition step ?

I decided to change data preprocess step like this now : data x =
cat(node_feature, edge_attributes) so the concate is done in the data.

But I have still an issue on the neighbour feature weighting gate part.
Based on this paper, we need to learn the weights per feature not per
atoms.

I don't know how to make this operation in pytorch / PyG indeed.

that would be nice if you can take a look please,

thank you

Best regards,

Guillaume



Le lun. 19 oct. 2020 à 10:27, Matthias Fey <notifications@github.com> a
écrit :

**Comment by rusty1s** [comment]
Yes, Eq (7) looks wrong from a dimensionality perspective. You can either append zero vectors to your central node features or use a linear projection to transform incoming messages to the same dimensionality.

IMO, you already learn the weights per feature since `gate_x_j` computes a weight for each feature.
